File: javascript/workflow/src/verbs/groupby.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { GroupbyArgs } from '@datashaper/schema'

import type { ColumnTableStep } from './util/factories.js'
import { stepVerbFactory } from './util/factories.js'

export const groupbyStep: ColumnTableStep<GroupbyArgs> = (input, { columns }) =>
	input.groupby(columns)

export const groupby = stepVerbFactory(groupbyStep)

File: python/datashaper/datashaper/verbs/groupby.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Groupby verb implementation."""
import pandas as pd
from pandas.core.groupby import DataFrameGroupBy
from reactivedataflow import ConfigPort, InputPort, verb

from datashaper import DEFAULT_INPUT_NAME

from .decorators import OutputMode, copy_input_tables, wrap_verb_result


@verb(
    name="groupby",
    ports=[
        InputPort(name=DEFAULT_INPUT_NAME, parameter="table", required=True),
        ConfigPort(name="columns", required=True),
    ],
    adapters=[
        copy_input_tables("table"),
        wrap_verb_result(mode=OutputMode.Table),
    ],
)
def groupby(table: pd.DataFrame, columns: list[str]) -> DataFrameGroupBy:
    """Groupby verb implementation."""
    return table.groupby(by=columns)
